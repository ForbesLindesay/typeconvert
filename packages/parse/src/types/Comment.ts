import {SourceLocation} from '@typeconvert/file-context';

export enum CommentKind {
  CommentBlock = 'CommentBlock',
  CommentLine = 'CommentLine',
}

export interface Comment {
  readonly kind: CommentKind;
  readonly value: string;
  readonly loc: SourceLocation;
}
export default Comment;
