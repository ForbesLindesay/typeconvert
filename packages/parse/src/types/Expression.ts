import ArrayExpresison from './ExpressionTypes/ArrayExpresison';
import Identifier from './ExpressionTypes/Identifier';

type Expression = ArrayExpresison | Identifier;
export default Expression;
