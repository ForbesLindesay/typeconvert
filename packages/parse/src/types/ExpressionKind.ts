enum ExpressionKind {
  ArrayExpression = 'ArrayExpression',
  Identifier = 'Identifier',
}
export default ExpressionKind;
