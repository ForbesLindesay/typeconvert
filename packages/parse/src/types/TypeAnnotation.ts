import AnyTypeAnnotation from './TypeAnnotationTypes/AnyTypeAnnotation';
import ArrayTypeAnnotation from './TypeAnnotationTypes/ArrayTypeAnnotation';
import BooleanTypeAnnotation from './TypeAnnotationTypes/BooleanTypeAnnotation';
import LiteralTypeAnnotation from './TypeAnnotationTypes/LiteralTypeAnnotation';
import ExistsTypeAnnotation from './TypeAnnotationTypes/ExistsTypeAnnotation';
import FunctionTypeAnnotation from './TypeAnnotationTypes/FunctionTypeAnnotation';
import TypeReferenceAnnotation from './TypeAnnotationTypes/TypeReferenceAnnotation';
import IntersectionTypeAnnotation from './TypeAnnotationTypes/IntersectionTypeAnnotation';
import UnknownTypeAnnotation from './TypeAnnotationTypes/UnknownTypeAnnotation';
import EmptyTypeAnnotation from './TypeAnnotationTypes/EmptyTypeAnnotation';
import NumberTypeAnnotation from './TypeAnnotationTypes/NumberTypeAnnotation';
import StringTypeAnnotation from './TypeAnnotationTypes/StringTypeAnnotation';
import TupleTypeAnnotation from './TypeAnnotationTypes/TupleTypeAnnotation';
import ThisTypeAnnotation from './TypeAnnotationTypes/ThisTypeAnnotation';
import InferTypeAnnotation from './TypeAnnotationTypes/InferTypeAnnotation';
import UnionTypeAnnotation from './TypeAnnotationTypes/UnionTypeAnnotation';
import TypeofTypeAnnotation from './TypeAnnotationTypes/TypeofTypeAnnotation';
import ImportTypeAnnotation from './TypeAnnotationTypes/ImportTypeAnnotation';
import ObjectTypeAnnotation from './TypeAnnotationTypes/ObjectTypeAnnotation';
import SymbolTypeAnnotation from './TypeAnnotationTypes/SymbolTypeAnnotation';

type TypeAnnotation =
  | AnyTypeAnnotation
  | ArrayTypeAnnotation
  | BooleanTypeAnnotation
  | LiteralTypeAnnotation
  | ExistsTypeAnnotation
  | FunctionTypeAnnotation
  | TypeReferenceAnnotation
  // | InterfaceTypeAnnotation
  | IntersectionTypeAnnotation
  | UnknownTypeAnnotation
  | EmptyTypeAnnotation
  // | NullableTypeAnnotation
  | NumberTypeAnnotation
  | ObjectTypeAnnotation
  | SymbolTypeAnnotation
  | StringTypeAnnotation
  | ThisTypeAnnotation
  | TupleTypeAnnotation
  // | TypeofTypeAnnotation
  | UnionTypeAnnotation
  // | VoidTypeAnnotation
  | InferTypeAnnotation
  | TypeofTypeAnnotation
  | ImportTypeAnnotation;
export default TypeAnnotation;
