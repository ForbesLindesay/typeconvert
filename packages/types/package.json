{
  "name": "@typeconvert/types",
  "version": "0.0.0",
  "description": "",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "dependencies": {
    "@babel/types": "^7.4.0",
    "@types/node": "^8.0.33"
  },
  "scripts": {
    "build": "tsc -p tsconfig.build.json && node ../../scripts/prepare types"
  },
  "repository": "https://github.com/ForbesLindesay/typeconvert/tree/master/packages/types",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "devDependencies": {
    "typescript": "^3.3.4000"
  }
}
