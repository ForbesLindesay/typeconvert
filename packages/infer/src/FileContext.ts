import * as bt from '@babel/types';
import {Mode} from '@typeconvert/types';
const {codeFrameColumns} = require('@babel/code-frame');

export default class FileContext {
  readonly filename: string;
  readonly src: string;
  readonly mode: Mode;

  constructor(filename: string, src: string, mode: Mode) {
    this.filename = filename;
    this.src = src;
    this.mode = mode;
  }

  getError(msg: string, node: {loc: bt.Node['loc']}) {
    const loc = node.loc;
    if (loc) {
      return new Error(
        msg +
          '\n\n' +
          codeFrameColumns(this.src, loc, {
            highlightCode: true,
          }),
      );
    } else {
      return new Error(msg);
    }
  }

  // N.B. convert this back to `never` to expose errors
  assertNever(msg: string, value: never): never {
    const v: any = value;
    if (typeof v === 'object' && typeof v.loc === 'object' && v.loc) {
      throw this.getError(msg, v);
    }
    throw new Error(msg + ' in ' + this.filename);
  }
}
