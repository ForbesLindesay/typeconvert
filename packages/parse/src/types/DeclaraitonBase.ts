import {SourceLocation} from '@typeconvert/file-context';
import Comment from './Comment';

export default interface DeclarationBase {
  readonly localName: string;
  readonly loc: SourceLocation | null;
  readonly leadingComments: ReadonlyArray<Comment> | null;
};
