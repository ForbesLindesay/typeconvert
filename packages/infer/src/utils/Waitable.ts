import * as bt from '@babel/types';
import FileContext from '../FileContext';
const {codeFrameColumns} = require('@babel/code-frame');

export type DeeperWaitable<T> =
    T extends any[] ? DeepWaitableArray<T[number]> :
    T extends ReadonlyArray<any> ? ReadonlyDeepWaitableArray<T[number]> :
    T extends object ? DeepWaitableObject<T> :
    T;

export interface DeepWaitableArray<T> extends Array<MaybeDeepWaitable<T>> {}
export interface ReadonlyDeepWaitableArray<T> extends ReadonlyArray<MaybeDeepWaitable<T>> {}

export type DeepWaitableObject<T> = {
    [P in keyof T]: P extends 'kind' | 'loc' ? T[P] : MaybeDeepWaitable<T[P]>;
};

export type UnWaitable<T> = 
  T extends Waitable<infer R> ? DeepUnwaitable<R>
  : DeepUnwaitable<T>;
export type DeepUnwaitable<T> =
    T extends any[] ? UnWaitableArray<T[number]> :
    T extends ReadonlyArray<any> ? UnWaitableReadonlyArray<T[number]> :
    T extends object ? UnWaitableObject<T> :
    T;
export interface UnWaitableArray<T> extends Array<UnWaitable<T>> {}
export interface UnWaitableReadonlyArray<T> extends ReadonlyArray<UnWaitable<T>> {}
export type UnWaitableObject<T> = {
  [P in keyof T]: UnWaitable<T[P]>;
}

enum Status {
  Pending,
  Fulfilled,
  Rejected,
}
type State<T> = 
  | {status: Status.Pending}
  | {status: Status.Fulfilled, value: T}
  | {status: Status.Rejected, error: any};

export type MaybeDeepWaitable<T> = DeeperWaitable<T> | Waitable<DeeperWaitable<T>>;
export type DeepWaitable<T> = Waitable<DeeperWaitable<T>>;

class Waitable<T> {
  private readonly _loc: bt.Node['loc'] | undefined;
  private readonly _ctx: FileContext | undefined;
  private _getValue: (() => T) | undefined;
  private _state: State<T> = {status: Status.Pending};
  constructor(getValue: () => T, loc?: bt.Node['loc'], ctx?: FileContext) {
    this._getValue = getValue;
    this._loc = loc;
    this._ctx = ctx;
  }
  getValue(): T {
    switch (this._state.status) {
      case Status.Pending:
        try {
          const getValue = this._getValue;
          this._getValue = undefined;
          if (!getValue) {
            const err = new Error('Unable to resolve cyclic dependency');
            (err as any)._waitableCycle = [];
            throw err;
          }
          const value = getValue();
          this._state = {status: Status.Fulfilled, value};
          return value;
        } catch (error) {
          const waitableCycle: Waitable<any>[] | undefined = (error as any)._waitableCycle;
          if (waitableCycle) {
            if (waitableCycle.indexOf(this) !== -1) {
              (error as any)._waitableCycle = undefined;
            } else {
              waitableCycle.push(this);
              if (this._loc && this._ctx) {
                error.message += '\n\n' + codeFrameColumns(this._ctx.src, this._loc, {
                  highlightCode: Waitable.highlightCodeInErrors,
                });
              }
            }
          }
          this._state = {status: Status.Rejected, error};
          throw error;
        }
      case Status.Fulfilled:
        return this._state.value;
      case Status.Rejected:
        throw this._state.error;
    }
  }
  getValueDeep(): UnWaitable<T> {
    return Waitable.resolveDeep(this.getValue());
  }
  then<S>(fn: (value: T) => Waitable<S> | S): Waitable<S> {
    return new Waitable((): S => {
      const result = fn(this.getValue());
      if (result instanceof Waitable) {
        return result.getValue();
      } else {
        return result;
      }
    });
  }
  thenDeep<S>(fn: (value: UnWaitable<T>) => Waitable<S> | S): Waitable<S> {
    return new Waitable((): S => {
      const result = fn(this.getValueDeep());
      if (result instanceof Waitable) {
        return result.getValue();
      } else {
        return result;
      }
    });
  }

  static highlightCodeInErrors: boolean = true;
  static resolve<T>(value: Waitable<T> | T): T {
    if (value instanceof Waitable) {
      return value.getValue();
    } else {
      return value;
    }
  }
  static resolveDeep<T>(value: T): UnWaitable<T> {
    // the slightly odd data structure is to be able to handle recursive structures:
    // e.g.
    // const x = new Waitable(() => 10);
    // const y = {x};
    // y.y = y;
    // const z = new Waitable(() => y);
    // Waitable.resolveDeep(z);
    const cache = new Map<any, {completed: true, value: any} | {completed: false, pending: Array<(value: any) => any>}>();
    function recurse(value: any, onResult: (value: any) => void) {
      const cached = cache.get(value);
      if (cached) {
        if (cached.completed) {
          onResult(cached.value);
        } else {
          cached.pending.push(onResult);
        }
        return;
      }
      const pending = [onResult];
      cache.set(value, {completed: false, pending});
      onResult = (result) => {
        cache.set(value, {completed: true, value: result});
        pending.forEach(p => p(result));
      };
      if (value instanceof Waitable) {
        recurse(value.getValue(), onResult);
        return;
      }
      if (Array.isArray(value)) {
        const result: any[] = new Array(value.length);
        value.forEach((value, i) => {
          recurse(value, value => result[i] = value);
        })
        onResult(result);
        return;
      }
      if (value && typeof value === 'object') {
        const result: any = {};
        for (const key in value) {
          recurse(value[key], value => result[key] = value);
        }
        onResult(result);
        return;
      }
      onResult(value);
    }
    let result: any;
    recurse(value, v => result = v);
    return result;
  }
}
export default Waitable;