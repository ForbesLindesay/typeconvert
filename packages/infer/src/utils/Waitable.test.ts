import {readFileSync} from 'fs';
import {Mode} from '@typeconvert/types';
import Waitable from './Waitable';
import FileContext from '../FileContext';

test('Waitable.resolveDeep - circular', () => {
  const x = new Waitable(() => 10);
  const y: any = {x};
  y.y = y;
  const z = new Waitable(() => y);
  const expected: any = {x: 10};
  expected.y = expected;
  expect(Waitable.resolveDeep(z)).toEqual(expected);
});

test('Waitable.resolveDeep - types', () => {
  const x = new Waitable(() => 10);
  const y = {x};
  const z = new Waitable(() => y);
  const expected = {x: 10};
  const result: {x: number} = Waitable.resolveDeep(z);
  expect(result).toEqual(expected);
});

test('waitable.getValue() -> cyclic dependency error', () => {
  const x = new Waitable(
    (): number => y.getValue(),
    {
      start: {
        line: 1,
        column: 0,
      },

      end: {
        line: 1,
        column: 10,
      },
    },
    new FileContext(
      __filename,
      readFileSync(__filename, 'utf8'),
      Mode.typescript,
    ),
  );
  const y = new Waitable(
    (): number => z.getValue(),
    {
      start: {
        line: 2,
        column: 0,
      },

      end: {
        line: 2,
        column: 10,
      },
    },
    new FileContext(
      __filename,
      readFileSync(__filename, 'utf8'),
      Mode.typescript,
    ),
  );
  const z = new Waitable(
    (): number => x.getValue(),
    {
      start: {
        line: 3,
        column: 0,
      },

      end: {
        line: 3,
        column: 10,
      },
    },
    new FileContext(
      __filename,
      readFileSync(__filename, 'utf8'),
      Mode.typescript,
    ),
  );
  const a = new Waitable(
    (): number => x.getValue(),
    {
      start: {
        line: 42,
        column: 0,
      },

      end: {
        line: 42,
        column: 10,
      },
    },
    new FileContext(
      __filename,
      readFileSync(__filename, 'utf8'),
      Mode.typescript,
    ),
  );
  Waitable.highlightCodeInErrors = false;
  expect(() => a.getValue()).toThrowErrorMatchingSnapshot();
  expect(() => a.getValue()).toThrowErrorMatchingSnapshot();
  Waitable.highlightCodeInErrors = true;
});
