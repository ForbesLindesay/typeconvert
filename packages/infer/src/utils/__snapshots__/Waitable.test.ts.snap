// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`waitable.getValue() -> cyclic dependency error 1`] = `
"Unable to resolve cyclic dependency

> 1 | import {readFileSync} from 'fs';
    | ^^^^^^^^^^
  2 | import {Mode} from '@typeconvert/types';
  3 | import Waitable from './Waitable';
  4 | import FileContext from '../FileContext';

  1 | import {readFileSync} from 'fs';
  2 | import {Mode} from '@typeconvert/types';
> 3 | import Waitable from './Waitable';
    | ^^^^^^^^^^
  4 | import FileContext from '../FileContext';
  5 | 
  6 | test('Waitable.resolveDeep - circular', () => {

  1 | import {readFileSync} from 'fs';
> 2 | import {Mode} from '@typeconvert/types';
    | ^^^^^^^^^^
  3 | import Waitable from './Waitable';
  4 | import FileContext from '../FileContext';
  5 | "
`;

exports[`waitable.getValue() -> cyclic dependency error 2`] = `
"Unable to resolve cyclic dependency

> 1 | import {readFileSync} from 'fs';
    | ^^^^^^^^^^
  2 | import {Mode} from '@typeconvert/types';
  3 | import Waitable from './Waitable';
  4 | import FileContext from '../FileContext';

  1 | import {readFileSync} from 'fs';
  2 | import {Mode} from '@typeconvert/types';
> 3 | import Waitable from './Waitable';
    | ^^^^^^^^^^
  4 | import FileContext from '../FileContext';
  5 | 
  6 | test('Waitable.resolveDeep - circular', () => {

  1 | import {readFileSync} from 'fs';
> 2 | import {Mode} from '@typeconvert/types';
    | ^^^^^^^^^^
  3 | import Waitable from './Waitable';
  4 | import FileContext from '../FileContext';
  5 | "
`;
